### PR TITLE
Add support for unique expression indexes

### PR DESCRIPTION
**Pull request by matobet**
Fixes the problem where a index in the form of

```
  uniqueIndexOnColumns
    [ "role"
    , "src_id"
    , "(src_id IS NULL)"
    , "trg_id"
    , "(trg_id IS NULL)"
    ]
```

would cause the problem when comparing the created index name with expected:
`expected: unique_idx__roles__role__src_id__$src_id$IS$NULL$__trg_id__$trg`
`given:       unique_idx__roles__role__src_id__$src_id$is$null$__trg_id__$trg`
and failing the respective migration.

Now we lowercase all chars in `indexName` normalizing the index name.

https://reviewable.io/reviews/scrive/hpqtypes-extras/35

### CODE REVIEW

**Comment by EggBaconAndSpam** [comment]
Do we know why hpqtypes ends up capitalising the `IS` in the first place?

**Comment by matobet** [comment]
For reference, changing the attached test case to lowercase `is null` causes the following error:
```
  Test unique partial expression indexes: 2020-12-02 15:22:40 INFO hpqtypes-extras-test: Creating domains...
2020-12-02 15:22:40 INFO hpqtypes-extras-test: Creating tables...
2020-12-02 15:22:40 INFO hpqtypes-extras-test: Creating table constraints...
2020-12-02 15:22:40 INFO hpqtypes-extras-test: Done.
2020-12-02 15:22:40 INFO hpqtypes-extras-test: Running initial setup for tables...
2020-12-02 15:22:40 INFO hpqtypes-extras-test: Done.
2020-12-02 15:22:40 ATTENTION hpqtypes-extras-test: There are problems with the table 'roles'
2020-12-02 15:22:40 ATTENTION hpqtypes-extras-test: Table and its definition have diverged and have 1 and 1 different INDEXes each, respectively (table: [TableIndex {idxColumns = [RawSQL "role" (),RawSQL "src_id" (),RawSQL "(src_id IS NULL)" (),RawSQL "trg_id" (),RawSQL "(trg_id IS NULL)" ()], idxMethod = btree, idxUnique = True, idxValid = True, idxWhere = Nothing}], definition: [TableIndex {idxColumns = [RawSQL "role" (),RawSQL "src_id" (),RawSQL "(src_id is null)" (),RawSQL "trg_id" (),RawSQL "(trg_id is null)" ()], idxMethod = btree, idxUnique = True, idxValid = True, idxWhere = Nothing}]).
FAIL (0.02s)
    Dropping the test DB schema... (0.02s)
      resultCheck: validation failed
      CallStack (from HasCallStack):
        error, called at src/Database/PostgreSQL/PQTypes/Checks/Util.hs:85:7 in hpqtypes-extras-1.10.3.0-inplace:Database.PostgreSQL.PQTypes.Checks.Util

1 out of 5 tests failed (1.26s)
Test suite hpqtypes-extras-tests: FAIL
Test suite logged to:
/home/matobet/work/scrive/hpqtypes-extras/dist-newstyle/build/x86_64-linux/ghc-8.8.4/hpqtypes-extras-1.10.3.0/t/hpqtypes-extras-tests/test/hpqtypes-extras-1.10.3.0-hpqtypes-extras-tests.log
0 of 1 test suites (0 of 1 test cases) passed.
cabal: Tests failed for test:hpqtypes-extras-tests from
hpqtypes-extras-1.10.3.0.
```

**Comment by matobet** [comment]
with regards to PostgresSQL behavior it seems to lowercase all names (try `CREATE idx_MY_DUMMY` and you will see it will create `"idx_my_dummy"` so with this I believe having the `toLower` in that place is the correct thing to do irrespective.

**Comment by EggBaconAndSpam** [comment]
It appears the postgres lower-cases all identifiers by default; from the [postgres manual](https://www.postgresql.org/docs/current/sql-syntax-lexical.html):

> Quoting an identifier also makes it case-sensitive, whereas unquoted names are always folded to lower case. For example, the identifiers FOO, foo, and "foo" are considered the same by PostgreSQL, but "Foo" and "FOO" are different from these three and each other. (The folding of unquoted names to lower case in PostgreSQL is incompatible with the SQL standard, which says that unquoted names should be folded to upper case. Thus, foo should be equivalent to "FOO" not "foo" according to the standard. If you want to write portable applications you are advised to always quote a particular name or never quote it.)

**Comment by jonathanjouty** [comment]
But dollar quoted strings _are_ case sensitive, however we're not lowercasing those I believe.

**Comment by zlondrej** [comment]
I believe that better solution would be to add double quotes around index name in `src/Database/PostgreSQL/PQTypes/Model/Index.hs`. Postgres lowercases the index name unless it's double quoted (mentioned in comment by @EggBaconAndSpam).
> Quoting an identifier also makes it case-sensitive

```diff
diff --git a/src/Database/PostgreSQL/PQTypes/Model/Index.hs b/src/Database/PostgreSQL/PQTypes/Model/Index.hs
index ca4a9cd..9637f68 100644
--- a/src/Database/PostgreSQL/PQTypes/Model/Index.hs
+++ b/src/Database/PostgreSQL/PQTypes/Model/Index.hs
@@ -154,7 +154,7 @@ sqlCreateIndex_ concurrently tname idx@TableIndex{..} = mconcat [
   , if idxUnique then " UNIQUE" else ""
   , " INDEX "
   , if concurrently then "CONCURRENTLY " else ""
-  , indexName tname idx
+  , "\"" <> indexName tname idx <> "\""
   , " ON" <+> tname
   , " USING" <+> (rawSQL (T.pack . show $ idxMethod) ()) <+> "("
   , mintercalate ", " idxColumns
```

You might also want to bump version in `cabal` file. Although more correct approach would be to do that right before release.

**Comment by arybczak** [comment]
It looks to me that if you create the index as follows:

```
  uniqueIndexOnColumns
    [ "role"
    , "src_id"
    , "(src_id is null)"
    , "trg_id"
    , "(trg_id is null)"
    ]
```

then it will work fine without this PR.

**Comment by EggBaconAndSpam** [comment]
@arybczak In that case you'll get a constraint mismatch, since postgres capitalises keywords (i.e. turn `is null` into `IS NULL`).

**Comment by arybczak** [comment]
Oh, ok.

In that case @zlondrej's solution seems to be more principled :thinking:

**Comment by matobet** [comment]
@arybczak @zlondrej I agree, escaping is much more general and more correct. Done.

File: src/Database/PostgreSQL/PQTypes/Model/Index.hs [hunk 1]
```diff
@@ -154,9 +154,9 @@ sqlCreateIndex_ concurrently tname idx@TableIndex{..} = mconcat [
   , if idxUnique then " UNIQUE" else ""
   , " INDEX "
   , if concurrently then "CONCURRENTLY " else ""
-  , indexName tname idx
+  , "\"", indexName tname idx, "\""
```
Comment (zlondrej) [inline src/Database/PostgreSQL/PQTypes/Model/Index.hs:157]: Turns out that you'll have quote every occurence of `indexName` that makes it's way into SQL query.
I suggest to make quoting function like `dquoted`.

Same change might be also appropriate in `PrimaryKey.hs` and `ForeignKey.hs`.

File: test/Main.hs [hunk 2]
```diff
@@ -906,6 +906,54 @@ migrationTest4 connSource =
...
+                    , tblColumns     =
+                      [ tblColumn { colName = "id", colType = BigSerialT, colNullable = False }
+                      , tblColumn { colName = "role", colType = SmallIntT, colNullable = False }
+                      , tblColumn { colName = "src_id", colType = BigIntT, colNullable = True }
+                      , tblColumn { colName = "trg_id", colType = BigIntT, colNullable = True }
+                      ]
+                    , tblPrimaryKey  = pkOnColumn "id"
+                    , tblIndexes     = [indexOnColumn "src_id"
+                                      , indexOnColumn "trg_id"
+                                      , uniqueIndexOnColumns
```
Comment (zlondrej) [inline test/Main.hs:927]: Also test `sqlDropIndex`.

**Review by zlondrej — CHANGES_REQUESTED** [review]

File: src/Database/PostgreSQL/PQTypes/Model/Index.hs (hunk 1 above)
Comment (matobet) [reply src/Database/PostgreSQL/PQTypes/Model/Index.hs]: Done.

**Review by zlondrej — CHANGES_REQUESTED** [review]
Your tests are still failing. Apparently you have to parenthesize this expression `COALESCE(trg_id, - 1::bigint)`.

I have no other comments.

**Comment by matobet** [comment]
This is really strange... the test are passing for me locally. Will have to investigate.

**Comment by matobet** [comment]
@zlondrej fixed, added the parentheses, now the tests pass on Travis but fail for me locally... Not sure if this is caused by the postgres version (11.10 for me) or what, but w.r.t. to this PR this appears to be fixed....

**Comment by zlondrej** [comment]
I'm more surprised that it passed for you before. As Postgres sytax for `CREATE INDEX` shows that index expression has to be parenthesized.
```
CREATE [ UNIQUE ] INDEX [ CONCURRENTLY ] [ [ IF NOT EXISTS ] name ] ON [ ONLY ] table_name [ USING method ]
    ( { column_name | ( expression ) } [ COLLATE collation ] [ opclass [ ( opclass_parameter = value [, ... ] ) ] ] [ ASC | DESC ] [ NULLS { FIRST | LAST } ] [, ...] )
                      ^ ---------- ^ <- expression has to be parenthesized
```

Another example is here: https://www.postgresql.org/docs/current/indexes-expressional.html

Anyway, this discrepancy doesn't look like big issue as you just have to adjust expression based on your database.
For example there's `checkChecks` function in `src/Database/PostgreSQL/PQTypes/Checks.hs` which has error message with hint:
> HINT: If checks are equal modulo number of parentheses/whitespaces used in conditions, just copy and paste expected output into source code
Something similar could be added to `checkIndexes`.

**Review by zlondrej — APPROVED** [review]

**Comment by arybczak** [comment]
I feel uneasy about the fact that you need to insert `quote` everywhere. Once you forget it somewhere you'll have a bug. Can this be improved?

**Comment by matobet** [comment]
Yeah, the thing is that it should be "everywhere" (otherwise we could do push it into the `indexName` function), just on the places where you concatenate SQL. For example `checkNames` (`checkIndexes`) relies on the unquoted version.

**Comment by matobet** [comment]
Alternatively we could have the *name functions return not `RawSQL` but some "Identifier" newtype wrapper that would have to be converted - either to `RawSQL` (by quoting it) or to `Text` for comparison purposes. This way the type system would force us to quote the identifiers where necessary.

**Comment by zlondrej** [comment]
I'm on mobile so my suggestion will be simplified.
You could make `indexName` (and others) return wrapper type that has to be unpacked in check function and quoted in sql expression. I don't know if `IsSQL` instance could be used for that.

**Comment by matobet** [comment]
@arybczak @zlondrej please review new version with `SqlIdentifier` newtype wrapper.

**Comment by arybczak** [comment]
Sorry for postponing a review but I'd like to have a new version with only no-downtime compatible changes to narrow possibility for issues. I'll get to it after it's done.

**Comment by matobet** [comment]
@arybczak ping, are we OK to merge this as it is?

**Comment by arybczak** [comment]
Sorry, I managed to completely ignore this PR since there is no use case for this (afaik?). I'll try to get to it when I have a spare moment.

**Comment by matobet** [comment]
Updated with latest `master`

**Comment by arybczak** [comment]
I'm gonna close this since we have no need for it. If it arises in the future, we can use this PR as a base.

**Pull request closed without merging**